Completely rewrite contributing guidelines (#875)
ProcursusTeam/Procursus: CONTRIBUTING.md

# Contributing to Procursus

> Community involvement is one of the main goals of this project, so we're glad you'd like to make a contribution. There are a few requirements, but they shouldn't be too unresonable.

Reviewer: ```suggestion
Community involvement is one of the main goals of this project, so we're glad you'd like to make a contribution. There are a few requirements, but they shouldn't be too unreasonable.
```

> If you have any questions, or would like to contact team members, you can either

Reviewer: ```suggestion
If you have any questions, or would like to contact team members, you can either:
```

Author: This isn't needed.

> - Email [Hayden](mailto:me@diatr.us) about your issue/concern (or annoy him on [Twitter](https://twitter.com/Diatrus)), or..

Reviewer: ```suggestion
- Email [Hayden](mailto:me@diatr.us) about your issue/concern (or annoy him on [Twitter](https://twitter.com/Diatrus)), or...
```

Author: No, I'm keeping it my way.

> - Email [Hayden](mailto:me@diatr.us) about your issue/concern (or annoy him on [Twitter](https://twitter.com/Diatrus)), or..

Reviewer: ```suggestion
- Email [Hayden](mailto:me@diatr.us) about your issue/concern (or annoy him on [Twitter](https://twitter.com/Diatrus))
```

> - Join [Hayden's Discord Server](https://diatr.us/discord), where most team members are capable of answering questions and provide support

Reviewer: ```suggestion
- Join [Hayden's Discord Server](https://diatr.us/discord), where most team members are capable of answering questions and providing support
```

## Submitting Issues

> Any issue with the buildsystem, or any package on the APT repo can be submitted as a Github Issue, and will be tracked as soon as possible.

Reviewer: ```suggestion
Any issues with the build system or a package on the APT repo can be submitted as a GitHub issue, and will be tracked as soon as possible.
```

> While there isn't an overbaring policy on how to properly submit issues, please be courteous and provide as much information about the situation as possible **in a cohesive matter.** Low effort or low info issues will be closed and marked as invalid.

Reviewer: ```suggestion
While there isn't an overbearing policy on how to properly submit issues, please be courteous and provide as much information about the situation as possible **in a cohesive matter.** Low effort or low info issues will be closed and marked as invalid.
```

## Contribution Process
1. Fork the project and make changes relevant to what you're trying to add/fix — do this in a seperate branch in your fork
2. Test your changes made on a **physical** device. If you're adding or updating a package, test your changes with a build of the package you added/updated
3. Create a Pull Request for your changes in your specific branch — include a description of the change and where you tested your change

> Don't expect your changes to be merged to pushed to upstream immediately — your changes will be reviewed by a team member (or Hayden), providing suggested changes which must be met.

Reviewer: ```suggestion
Don't expect your changes to be merged and pushed to upstream immediately — your changes will be reviewed by a team member (or Hayden), providing suggested changes which must be met.
```

Once all further changes have been met (if any) and your changes are approved, they will be merged into upstream. The APT repository will shortly update showcasing your changes.

### Adding a New Package

> Adding new packages to Procursus is fairly easy and simple. For most packages, you're likely to need 2 files

Reviewer: ```suggestion
Adding new packages to Procursus is fairly easy and simple. For most packages, you're likely to need 2 files:
```

Author: It's fine my way.

> Adding new packages to Procursus is fairly easy and simple. For most packages, you're likely to need 2 files

Reviewer 2: Please mention `./build_tools/new_package.sh`

Author: Done in latest commit.

> - A ``.mk`` file, that builds the package, and..

Reviewer: ```suggestion
- A `.mk` file that builds the package
```

> - A ``.control`` file, which contains info about your package

Reviewer: ```suggestion
- A `.control` file which contains info about your package
```

Some things to keep in mind when adding a new package

> - Always ensure that the Architecture and Maintainer fields of your control file are populated with ``@DEB_ARCH`` and ``@DEB_MAINTAINER@``, respectively

Reviewer: ```suggestion
- Always ensure that the Architecture and Maintainer fields of your control file are populated with `@DEB_ARCH@` and `@DEB_MAINTAINER@`, respectively
```

> - If you take advantage of already-made patchfiles from an external source, download them and implement them to your project. Check out how it's done in [``bash.mk``](./makefiles/bash.mk)

Reviewer: ```suggestion
- If you take advantage of already-made patch files from an external source, download them and implement them to your project. Check out how it's done in [``bash.mk``](./makefiles/bash.mk)
```

Author: Thanks, but it's fine the way I did it.

> - If the package you're adding only requires a few small edits, try taking advantage of ``sed`` in your packages' setup stage, and..

Reviewer: ```suggestion
- If the package you're adding only requires a few small edits, try taking advantage of ``sed`` in your package's setup stage
```

Author: "Packages's" translates to "package is"; the correct way is "packages'", since it's an attribute the package has, not what the package is.

Reviewer: Are you dumb? If you're referring to a single package, it's `package's`. `packages'` or `packages's` would be referring to *multiple* packages.

Author: My bad; I'll unresolve and agree with the changes.

- Use designated functions found in the [documentation](https://github.com/ProcursusTeam/Procursus/wiki) to download, patch, or make other changes to your package

### Updating a Package
Package update contributions are always welcome, as not all team members are aware of new versions provided by the build-system. For this specific scenario, there's two options

> - **Fork and pull request**. Most of the time, updating packages is as simple as updating the version number in the ``.mk`` file of the package and recompiling. If you have the ability to test new versions of packages, feel free to go this route (following [package addition guidelines](#adding-a-new-package)). Otherwise...

Reviewer: ```suggestion
- **Fork and pull request**. Most of the time, updating packages is as simple as updating the version number in the `.mk` file of the package and recompiling. If you have the ability to test new versions of packages, feel free to go this route (following [package addition guidelines](#adding-a-new-package)). Otherwise...
```

## Contribution Process
1. Fork the project, and make changes relevant to what you're trying to add/fix on a seperate branch
2. Create a Github Pull Request for your change in your specific branch. You should include a description of the change, why you're requesting your change to be made, and where you tested your change

> 3. Don't expect your changes to be merge to upstream instantly, since a team member will review your pull request, asking for further possible changes

Reviewer 3: Could this be worded better?

Author: Sure. I can suggest something

```
Don't expect your changes to be pushed to upstream automatically — your changes
must be reviewed by a team member, and suggested changes must be met before your pull request is approved.
```

Author: This will be changed in a new PR.
